pyanaconda/modules/boss/module_manager.py

```diff
@@ -0,0 +1,122 @@
+
+    @property
+    def modules_starting_finished(self):
+        return set(self._started_module_services + self._failed_module_services) == set(self.expected_module_services)
+
+    def _start_service_cb(self, service, returned=None, error=None):
+        """Callback for dbus.StartServiceByName."""
+        if error:
+            log.debug("%s error: %s", service, error)
+            self._failed_module_services.append(service)
+        elif returned:
+            if returned == dbus_constants.DBUS_START_REPLY_SUCCESS:
+                log.debug("%s started successfully, returned: %s)", service, returned)
+                self._started_module_services.append(service)
+            else:
+                log.debug("%s failed to start, returned: %s", service, returned)
+                self._failed_module_services.append(service)
+        else:
+            log.error("%s failed to start without even returning anything", service)
+
+        self.check_modules_started()
+
+    def check_modules_started(self):
+        if self.modules_starting_finished:
+            log.debug("modules starting finished, running: %s failed: %s",
+                      self._started_module_services,
+                      self._failed_module_services)
+            for service in self._started_module_services:
+                module = self._bus.get(service)
+                module.EchoString("Boss told me that all required modules were started: %s or failed: %s." %
+                                  (self._started_module_services, self._failed_module_services))
+            return True
+        else:
+            return False
+
+    def find_addons(self):
+        self._addon_module_services = []
+        names = self._dbus.ListActivatableNames()
+        for name in names:
+            if name.startswith(dbus_constants.DBUS_ADDON_NAMESPACE):
+                self._addon_module_services.append(name)
+
+    def start_modules(self):
+        """Starts anaconda modules (including addons)."""
+        log.debug("starting modules")
+
+        # Check that no modules are running
+        for service in self.expected_module_services:
+            if self._dbus.NameHasOwner(service):
+                log.error("service %s has unexpected owner", service)
+
+        for service in self.expected_module_services:
+            log.debug("Starting %s", service)
+            try:
+                self._dbus.StartServiceByName(service, 0, callback=self._start_service_cb, callback_args=(service,))
+            except Exception:  # pylint: disable=broad-except
+                self._failed_module_services.append(service)
+                log.exception("module startup failed")
+
+        log.debug("started all modules")
+
+    def stop_modules(self):
+        """Tells all running modules to quit."""
+        log.debug("sending Quit to all modules and addons")
+        for service in self.running_module_services:
+            module = self._bus.get(service)
+            # TODO: async ?
```

Comment:
I don't think it will be async at all. This should be run from the main thread as all the other modules related operations.

Reply:
It can be called from the main thread but still be async. Imagine a module hanging in Quit.

Reply:
> It can be called from the main thread but still be async. Imagine a module hanging in Quit.
> (Similar as we are starting modules asynchronously)
That's a really good point, definitely something to add later on. I'll add it to the TODO so it doesn't get lost.